### PR TITLE
dma: cAVS HDA DMA Driver

### PR DESCRIPTION
**Pull request by teburd**
Adds HDA host in and host out DMA drivers for Intel cAVS.

Each HDA stream direction is a driver here as the behavior needs to slightly vary. Most of the code is common code. Adding the link side drivers should be very easy.

A low level driver is available in the soc directory as a header, cavs_hda.h. This is meant to provide useful and direct access to manipulate registers. It enables unit testing with host side interaction without needing to map the hardware to the DMA API.

A DMA driver is built on top of the low level cavs_hda.h for host in and host out streams. In order to support the underlying ringbuffer the read_position and write_position were added to dma_status.

To inform the hardware of read or write changes to the underlying buffer, dma_reload may be called with the length of
the read/written bytes.

A quick mapping of how the DMA calls map to the existing SOF API would be...

From SOF to Zephyr...
 
dma_copy  ~= dma_reload
dma_status ~= dma_get_status

They aren't exactly the same, but very close.

### CODE REVIEW

File: soc/xtensa/intel_adsp/common/include/cavs-link.ld [hunk 1]
```diff
@@ -410,12 +410,13 @@ SECTIONS {
   . = SEGSTART_UNCACHED;
   _heap_sentry = .;
 
-  /* dma buffers */
-  .lpbuf (NOLOAD): {
+  /* dma buffers, require 128 byte alignment */
+  .hpbuf (NOLOAD): {
+    . = ALIGN(128);
```
Comment (andyross) [inline soc/xtensa/intel_adsp/common/include/cavs-link.ld:415]: Note that the modern SOF model is to allocate DMA memory from a unified heap, and the intent is that this will be extended via the multiheap model.  Putting buffers into static linker-managed sections is probably not what we want long term.  If you need to define a static buffer, just an "aligned(128)" attribute (or a soc-level macro to abstract that) is probably preferable.

I can't remember why the lpbuf section was left here.  IIRC there were some references to (but not uses of?) the section in the SOF source still, or something like that.

File: soc/xtensa/intel_adsp/common/include/cavs_hda.h [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+ * Provides low level calls to support cAVS HDA streams with
+ * minimal abstraction that allows testing the hardware
+ * and its demands separately from the intended DMA API
+ * usage. Notably all addresses, counts, and sizes
+ * are predefined here if not defined elsewhere to enable
+ * easier unit testing.
+ */
+
+#ifndef HDA_HOST_OUT_BASE
+#define HDA_HOST_OUT_BASE 0x72800
```
Comment (andyross) [inline soc/xtensa/intel_adsp/common/include/cavs_hda.h:26]: This belong in devicetree.  Also the ifndef pattern here seems a little suspicious.  Why would you be worried about multiple definitions?  Sounds like maybe a header is missing an include guard?

File: soc/xtensa/intel_adsp/common/include/cavs_hda.h [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+#define DGCS_BSC BIT(11) /* Buffer Segment Completion */
+#define DGCS_BOR BIT(10) /* Buffer Overrun */
+#define DGCS_BUR BIT(10) /* Buffer Underrun */
+#define DGCS_BF BIT(9) /* Buffer Full */
+#define DGCS_BNE BIT(8) /* Buffer Not Empty */
+#define DGCS_FIFORDY BIT(5) /* Enable FIFO */
+#define DGCS_BSCIE BIT(3) /* Buffer Segment Completion Interrupt Enable */
+
+/* Gateway Buffer Base Address */
+#define DGBBA(base, stream) ((volatile uint32_t *)(HDA_ADDR(base, stream) + 0x04))
```
Comment (andyross) [inline soc/xtensa/intel_adsp/common/include/cavs_hda.h:69]: Style thing: the rest of cavs likes to use C structs to define register blocks (a pattern I personally love, but will admit I didn't introduce).  Code in the rest of Zephyr tends to use offsets like this but with the sys_read/write() APIs.  Adding another scheme here ("macros returning pointer") is maybe a little confusing?

File: soc/xtensa/intel_adsp/common/include/cavs_hda.h [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+ * This can be invaluable when finding out why HDA isn't doing (or maybe is)
+ * doing what you want it to do. Macro so you get the file and line
+ * of the call site included.
+ *
+ * @param name String that contains a name of the hda stream (or anything really)
+ * @param base Base address of the IP register block
+ * @param sid Stream ID
+ */
+#define cavs_hda_dbg(name, base, sid)					\
+		printk("%s:%u %s(%u:0x%p), dgcs: 0x%x, dgbba 0x%x, "	\
```
Comment (andyross) [inline soc/xtensa/intel_adsp/common/include/cavs_hda.h:107]: Should probably be LOG_DBG().  Standard guidelines these days seem to be that printk() usage shouldn't be commited to new code at the driver level or above, it's for arch code, tests (some platforms don't have logging available) and ad-hoc debugging.  I actually like this rule because it makes for an easy way to grep to see that I haven't left any stray debug stuff in a PR.

Also: it's a macro, it should be CAVS_HDA_DBG() or CAVS_HDA_REG_DUMP() or something.  I think the all caps rule is actually a MISRA thing?

File: soc/xtensa/intel_adsp/common/include/cavs_hda.h [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+ * @param buf_size Buffer size in bytes Must be 128 byte aligned
+ *
+ * @retval -EBUSY if the HDA stream is already enabled
+ * @retval -EINVAL if the buf is not in L2, buf isn't aligned on 128 byte boundaries
+ * @retval 0 on Success
+ */
+static inline int cavs_hda_set_buffer(uint32_t base, uint32_t sid,
+				     uint8_t *buf, uint32_t buf_size)
+{
+	void *cached_buf = arch_xtensa_cached_ptr(buf);
```
Comment (andyross) [inline soc/xtensa/intel_adsp/common/include/cavs_hda.h:158]: I'm confused here: why does the address need to be in the cached mapping?  Does the hardware require this?  Per spec, the DMA layer should be upstream of the xtensa mapping and shouldn't care, no?   If it doesn't work to use the uncached address, we should have a comment here calling that out.

File: soc/xtensa/intel_adsp/common/include/cavs_hda.h [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+ * @param sid Stream ID
+ */
+static inline void cavs_hda_disable(uint32_t base, uint32_t sid)
+{
+	*DGCS(base, sid) &= ~(DGCS_GEN | DGCS_FIFORDY);
+}
+
+/**
+ * @brief Determine the number of unused bytes in the fifo
+ *
```
Comment (andyross) [inline soc/xtensa/intel_adsp/common/include/cavs_hda.h:217]: Pedantry, but you should clarify that this is only useful for an **input**[1] stream ID.  It returns the wrong sense of subtraction if you're trying to figure out how long until an stream inbound to the DSP fills up.


[1] HDA terminology.  "Input" from the host side, output from the DSP, of course.

File: soc/xtensa/intel_adsp/common/include/cavs_test.h [hunk 7]
```diff
@@ -0,0 +1,62 @@
+/* Copyright (c) 2022 Intel Corporation
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
Comment (andyross) [inline soc/xtensa/intel_adsp/common/include/cavs_test.h:3]: Nitpick: the motion of this header wants to be a separate patch.  Also this is really heavily coupled to the cavstool.py unit test script, so maybe a "cavstool.h" name or something might make it clearer what belongs here.  It's not just generic stuff for "testing".

File: soc/xtensa/intel_adsp/tools/cavstool.py [hunk 8]
```diff
@@ -18,11 +18,160 @@
 
 PAGESZ = 4096
 HUGEPAGESZ = 2 * 1024 * 1024
-HUGEPAGE_FILE = "/dev/hugepages/cavs-fw-dma.tmp"
+HUGEPAGE_FILE = "/dev/hugepages/cavs-fw-dma.tmp."
 
 # Log is in the fourth window, they appear in 128k regions starting at 512k
 WINSTREAM_OFFSET = (512 + (3 * 128)) * 1024
 
+class HDAStream:
```
Comment (andyross) [inline soc/xtensa/intel_adsp/tools/cavstool.py:26]: This looks like a perfect generalization of the stream from the limited output stream definition we had earlier.  But shouldn't we also be removing the older code from the loader and using this instead?

Also: this much surgery wants to be its own patch.  I'd suggest a refactoring that introduces this class and uses it instead of the older code, then dropping new protocols on top of it.

File: soc/xtensa/intel_adsp/tools/cavstool.py [hunk 9]
```diff
@@ -125,11 +276,14 @@ def map_phys_mem():
         tot = 1 + int(runx("awk '/HugePages_Total/ {print $2}' /proc/meminfo"))
         os.system(f"echo {tot} > /proc/sys/vm/nr_hugepages")
 
-    hugef = open(HUGEPAGE_FILE, "w+")
+    hugef_name = HUGEPAGE_FILE + str(stream_id)
+    hugef = open(hugef_name, "w+")
```
Comment (andyross) [inline soc/xtensa/intel_adsp/tools/cavstool.py:280]: Probably not a big deal, but note that these pages are MUCH larger than we need, so it's a little wasteful to have one per stream.  Longer term it might be worth trying to optimize usage.  But then... a few dozen MB isn't that much on a modern system.

File: soc/xtensa/intel_adsp/tools/cavstool.py [hunk 10]
```diff
@@ -125,11 +276,14 @@ def map_phys_mem():
...
     hugef.truncate(HUGEPAGESZ)
     mem = mmap.mmap(hugef.fileno(), HUGEPAGESZ)
+    log.info("type of mem is %s", str(type(mem)))
     global_mmaps.append(mem)
-    os.unlink(HUGEPAGE_FILE)
+    # helpful to sometimes comment out for debugging, can then cat the page
```
Comment (andyross) [inline soc/xtensa/intel_adsp/tools/cavstool.py:285]: Not sure that comment is all that helpful.  I mean, yes, that's the way hugetlbfs works, the man page says it better than we can here.  If you want a mode to leave the pages behind for debugging, maybe we should add that to the interface?

File: soc/xtensa/intel_adsp/tools/cavstool.py [hunk 11]
```diff
@@ -329,6 +484,45 @@ def ipc_command(data, ext_data):
...
+        hda_streams[stream_id].start()
+
+    elif data == 8: # HDA STOP
+        stream_id = ext_data & 0xFF
+        hda_streams[stream_id].stop()
+    elif data == 9: # HDA VALIDATE
+        stream_id = ext_data & 0xFF
+        hda_str = hda_streams[stream_id]
+        hda_str.debug()
+        is_ramp_data = True
```
Comment (andyross) [inline soc/xtensa/intel_adsp/tools/cavstool.py:510]: OK, you got me: I can't figure out what "ramp" means here. :)  Maybe just call this flag "valid" or something?

Also: maybe this is too much intelligence for the script?  If you just had a "read this offset from a stream buffer" command, you could do the validation in the C code where the buffer is generated, which is slower but IMHO a lot clearer and more flexible.

File: soc/xtensa/intel_adsp/tools/cavstool.py [hunk 12]
```diff
@@ -329,6 +484,45 @@ def ipc_command(data, ext_data):
...
+        hda_str.mem.seek(0)
+        for (i, val) in enumerate(hda_str.mem.read(256)):
+            if i != val:
+                is_ramp_data = False
+            log.info("stream[%d][%d]: %d", stream_id, i, val)
+        log.info("Is ramp data? " + str(is_ramp_data))
+        ext_data = int(is_ramp_data)
+        log.info(f"Ext data to send back on ramp status {ext_data}")
+        send_msg = True
+    elif data == 10: # HDA HOST OUT SEND
```
Comment (andyross) [inline soc/xtensa/intel_adsp/tools/cavstool.py:520]: Similarly, you could have a "write this word to this stream/offset" command and avoid the problem of having to synchronize the buffer formats between C and python.

File: soc/xtensa/intel_adsp/tools/cavstool.py [hunk 13]
```diff
@@ -343,7 +537,10 @@ def ipc_command(data, ext_data):
         dsp.HIPCIDR = (1<<31) | ext_data
 
 async def main():
-    global hda, sd, dsp, hda_ostream_id
+    log.debug("Removing old hugetables")
+    runx("rm -rf /dev/hugepages/cavs_*")
```
Comment (andyross) [inline soc/xtensa/intel_adsp/tools/cavstool.py:541]: What's this for?  The old scheme would detect and re-use the file if it was present.

File: soc/xtensa/intel_adsp/tools/cavstool.py [hunk 14]
```diff
@@ -343,7 +537,10 @@ def ipc_command(data, ext_data):
...
+    #TODO this bit me, remove the globals, write a little FirmwareLoader class or something to contain.
```
Comment (andyross) [inline soc/xtensa/intel_adsp/tools/cavstool.py:542]: Agreed.

File: tests/boards/intel_adsp/hda/src/smoke.c [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+#include <cavs_hda.h>
+#include "tests.h"
+
+#define IPC_TIMEOUT K_MSEC(100)
+#define STREAM_ID 3U
+#define FIFO_SIZE 256
+#define TRANSFER_SIZE 256
+#define TRANSFER_COUNT 8
+
+__attribute__((section(".dma_buffers"), aligned(128))) uint8_t hda_fifo[FIFO_SIZE];
```
Comment (andyross) [inline tests/boards/intel_adsp/hda/src/smoke.c:17]: As mentioned earlier, everything will work here if you drop the section and just make an aligned static buffer.

File: tests/boards/intel_adsp/hda/src/smoke.c [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+static bool ipc_message(const struct device *dev, void *arg,
+			uint32_t data, uint32_t ext_data)
+{
+	printk("HDA message received, data %u, ext_data %u\n", data, ext_data);
+	msg_res = data;
+	msg_cnt++;
+	return true;
+}
+
+static void ipc_done(const struct device *dev, void *arg)
```
Comment (andyross) [inline tests/boards/intel_adsp/hda/src/smoke.c:32]: It should work to register a null handler, you don't have to have an empty function.

File: tests/boards/intel_adsp/hda/src/smoke.c [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+	WAIT_FOR(cavs_ipc_send_message_sync(CAVS_HOST_DEV, IPCCMD_HDA_CONFIG,
+					    (STREAM_ID + 7) | (FIFO_SIZE << 8),
+					    IPC_TIMEOUT));
+	printk("host config: "); cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+
+	res = cavs_hda_set_buffer(HDA_HOST_OUT_BASE, STREAM_ID, hda_fifo, FIFO_SIZE);
+	printk("dsp set buffer: "); cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+	zassert_ok(res, "Expected set buffer to succeed");
+
+	k_sleep(K_MSEC(10));
```
Comment (andyross) [inline tests/boards/intel_adsp/hda/src/smoke.c:141]: What's this sleep for?

File: tests/boards/intel_adsp/hda/src/smoke.c [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+
+
+	for (uint32_t i = 0; i < TRANSFER_COUNT; i++) {
+
+		WAIT_FOR(cavs_ipc_send_message_sync(CAVS_HOST_DEV, IPCCMD_HDA_SEND,
+						    (STREAM_ID + 7) | (FIFO_SIZE << 8),
+						    IPC_TIMEOUT));
```
Comment (andyross) [inline tests/boards/intel_adsp/hda/src/smoke.c:148]: Not sure what's happening here?  This is going to iteratively retry the message send in a tight loop.  But it's a blocking call, it will wait on its own.   Why does it need a retry?  If you're getting spurious failures, there's likely some other bug.

Is it possible you send an async command earlier that hasn't completed yet?  If so, you want to poll on that and not the new message send.

File: tests/boards/intel_adsp/hda/src/smoke.c [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+	cavs_hda_disable(HDA_HOST_IN_BASE, STREAM_ID);
+}
+
+void test_hda_host_out_smoke(void)
+{
+	int res;
+	bool is_ramp;
+
+	printk("smoke testing hda with fifo buffer at address %p, size %d\n",
+	       hda_fifo, FIFO_SIZE);
```
Comment (andyross) [inline tests/boards/intel_adsp/hda/src/smoke.c:118]: Maybe a little too much printk traffic here for my taste.  In test code no one's going to read it unless it fails, so it's generally best to put the logging into the zassert_*() cases (which have a printf format string and arguments at the end).

File: tests/boards/intel_adsp/hda/src/smoke.c [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+	printk("host reset: "); cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+
+	WAIT_FOR(cavs_ipc_send_message_sync(CAVS_HOST_DEV, IPCCMD_HDA_CONFIG,
+					    (STREAM_ID + 7) | (FIFO_SIZE << 8),
+					    IPC_TIMEOUT));
+	printk("host config: "); cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+
+	res = cavs_hda_set_buffer(HDA_HOST_OUT_BASE, STREAM_ID, hda_fifo, FIFO_SIZE);
+	printk("dsp set buffer: "); cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+	zassert_ok(res, "Expected set buffer to succeed");
```
Comment (andyross) [inline tests/boards/intel_adsp/hda/src/smoke.c:139]: Shouldn't there be an assertion somewhere on the contents of the buffer, checking that the transfer moved the data correctly?

File: tests/boards/intel_adsp/hda/src/smoke.c [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+						    (STREAM_ID + 7), IPC_TIMEOUT));
+		printk("host start: ");
+		cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+
+		WAIT_FOR(cavs_hda_buf_full(HDA_HOST_OUT_BASE, STREAM_ID));
+		printk("dsp wait for full: ");
+		cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+
+		/* The buffer is in the cached address range and must be invalidated
+		 * prior to reading.
```
Comment (andyross) [inline tests/boards/intel_adsp/hda/src/smoke.c:166]: Uh... no it isn't.  Or if it is that's a linkage bug.  Static data needs to be uncached always.  You get caching on the stack, and on symbols with an explicit __incoherent tag applied (**and** where you've taken care to make sure they're correctly aligned and padded and won't overlap a cache line with anyone else).  That's the core metaphor behind KERNEL_COHERENCE.  You can rely on everything except local variables being coherent.

File: tests/boards/intel_adsp/hda/src/smoke.c [hunk 22]
```diff
@@ -0,0 +1,189 @@
...
+		 */
+		z_xtensa_cache_inv(hda_fifo, FIFO_SIZE);
+		is_ramp = true;
+		for (int j = 0; j < FIFO_SIZE; j++) {
+			printk("hda_fifo[%d] = %d\n", j, hda_fifo[j]);
```
Comment (andyross) [inline tests/boards/intel_adsp/hda/src/smoke.c:171]: Ah, OK.  Here's the buffer test.  Ignore my comment above.

Just a `zassert_equal(hda_fifo[j], j, "blah blah blah...");` is sufficient here.  That will longjmp out of the test for you on failure.  No need to track a flag through the whole iteration.

**Review by andyross — COMMENTED** [review]
Super quick read through.  Really all this looks great to me.  Very clean.

File: soc/xtensa/intel_adsp/common/include/cavs_hda.h (hunk 2 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/common/include/cavs_hda.h:26]: The intent here is yes, I'd move it to DTS when I go to implement the dma driver, its here because I needed something when testing. I can move this to my tests.

File: soc/xtensa/intel_adsp/common/include/cavs_hda.h (hunk 5 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/common/include/cavs_hda.h:158]: Its done so I can check the address is in L2  space which the hardware asks for. My interpretation may be wrong of the hardware docs but it does say cached or local L2. Local L2, I believe, is a different thing. Would the address range specified by cavs-mem L2_SRAM_BASE and L2_SRAM_SIZE be valid still if I chose uncached? That's my intent here, to ensure the address is in the range needed by the hardware to work.

File: soc/xtensa/intel_adsp/common/include/cavs_hda.h (hunk 4 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/common/include/cavs_hda.h:107]: Our style guide says "like linux" and linux has lower case macro functions all over the place. Kind of ugly to all cap this, but if its a MISRA thing I suppose its one of those things.

File: soc/xtensa/intel_adsp/common/include/cavs_hda.h (hunk 6 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/common/include/cavs_hda.h:217]: Definitely, in the host output side it'd be useful to see the number of used bytes instead. SOF combines the two into I think something like hda_usable_bytes() which I could also do here.

File: soc/xtensa/intel_adsp/tools/cavstool.py (hunk 8 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/tools/cavstool.py:26]: To be frank, I was frightened by the time it might take to do just that because of the seemingly endless hardware quirks. But it definitely tries to abstract it, and yes I should refactor the existing firmware load into this to some degree.

File: soc/xtensa/intel_adsp/tools/cavstool.py (hunk 9 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/tools/cavstool.py:280]: I considered writing a quick'n'dirty allocator on top of a single huge page for just that reason... but thought there were better uses of time, for the moment.

File: soc/xtensa/intel_adsp/tools/cavstool.py (hunk 13 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/tools/cavstool.py:541]: I wanted fresh and clean, every run. Which I didn't get without it. One of the many things I ended up falling into a rabbit hole on while debugging, probably removable now

File: tests/boards/intel_adsp/hda/src/smoke.c (hunk 15 above)
Comment (teburd) [reply tests/boards/intel_adsp/hda/src/smoke.c:17]: I'll try it, I had many issues, some of this stuff is left over because at the end of debugging I had a working version and wanted to move on to creating a dma driver on top of this. I tried reverting the linker script changes made above and the test failed so once it worked... I left well enough alone

File: tests/boards/intel_adsp/hda/src/smoke.c (hunk 16 above)
Comment (teburd) [reply tests/boards/intel_adsp/hda/src/smoke.c:32]: I had some debug printks in here before, removable now. Noted for the cleanup round

File: tests/boards/intel_adsp/hda/src/smoke.c (hunk 19 above)
Comment (teburd) [reply tests/boards/intel_adsp/hda/src/smoke.c:118]: For sure, was very helpful when I tested many, many times to see the screen blip by with my ramp data (0-255) or all zeros. Definitely overkill now that I have something working.

File: tests/boards/intel_adsp/hda/src/smoke.c (hunk 17 above)
Comment (teburd) [reply tests/boards/intel_adsp/hda/src/smoke.c:141]: Magic

File: tests/boards/intel_adsp/hda/src/smoke.c (hunk 21 above)
Comment (teburd) [reply tests/boards/intel_adsp/hda/src/smoke.c:166]: More leftover junk here, noted for the cleanup round

File: soc/xtensa/intel_adsp/common/include/cavs_hda.h (hunk 5 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/common/include/cavs_hda.h:158]: It's a 24 bit field which is fine, the goal in using the cached address is to validate the address is within the specified range and aligned appropriately. Fixed

File: soc/xtensa/intel_adsp/tools/cavstool.py (hunk 10 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/tools/cavstool.py:285]: Might be worth while adding a flag to leave the huge pages, I found it helpful when debugging

File: tests/boards/intel_adsp/hda/src/smoke.c (hunk 21 above)
Comment (teburd) [reply tests/boards/intel_adsp/hda/src/smoke.c:166]: Presumably I should be able to zassert_true(arch_mem_coherent(hda_fifo), "mem is coherent"); and that fails here for me

File: tests/boards/intel_adsp/hda/src/smoke.c (hunk 18 above)
Comment (teburd) [reply tests/boards/intel_adsp/hda/src/smoke.c:148]: I tried removing these, and the tests fail presumably because a message isn't getting sent. So maybe another bug to hunt down then.

File: tests/boards/intel_adsp/hda/src/smoke.c (hunk 22 above)
Comment (teburd) [reply tests/boards/intel_adsp/hda/src/smoke.c:171]: I want a long enough byte pattern, on both sides, that I can ensure things don't get corrupted. My initial test case actually sent a few bytes at a time in in a ring buffer with the data always containing the incremental byte values (0-255) in order, its pretty easy to spot if things have gone wrong, and no size limits by a predefined string then.

File: soc/xtensa/intel_adsp/common/include/cavs-link.ld (hunk 1 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/common/include/cavs-link.ld:415]: It was used in the s1000 I believe with the designware dma, could be wrong here though. It seems nice to have a dedicated linker section to dma buffers to me and it was already there when I got here, I just tweaked it to move where it was placed.

Certainly another, more integrated, test case might do a heap allocation, send over a loop with hda, and validate it end to end with the DMA API involved, we'll get there.

File: tests/boards/intel_adsp/hda/src/smoke.c (hunk 21 above)
Comment (teburd) [reply tests/boards/intel_adsp/hda/src/smoke.c:163]: It's because I had set MP_NUM_CPUS=1 in my test prj.conf trying to reduce what might be going wrong while debugging, this turns off the KERNEL_COHERENCE which is still probably wrong especially given the way HDA works.

**Comment by teburd** [comment]
A CI failure is going on caused by the previous gpdma work and each soc enabling DMA_DW without actually using it.

File: soc/xtensa/intel_adsp/common/include/cavs_hda.h (hunk 2 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/common/include/cavs_hda.h:23]: I've removed the ifdefs, to remove the defines would mean I should remove the low level driver in the header I think, which I found incredibly useful.

**Comment by teburd** [comment]
Working on enabling this for all soc's, I have a PR that I can base that on and rework this a little use it, see #43343

**Comment by teburd** [comment]
Now based on #43343 to avoid duplication of the device tree nodes and Kconfig defaults for intel_adsp family

File: soc/xtensa/intel_adsp/common/include/cavs_hda.h (hunk 3 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/common/include/cavs_hda.h:65]: This isn't the only driver to use a macro for a register returning a volatile pointer to the register. To be honest mapping this to structs, in my opinion, won't help build anymore clarity into the code here.

File: soc/xtensa/intel_adsp/common/include/cavs_hda.h (hunk 4 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/common/include/cavs_hda.h:103]: I looked through the MISRA docs, and our docs, and I didn't see anything that specifically says macros must be capped.

File: tests/boards/intel_adsp/hda/src/smoke.c (hunk 21 above)
Comment (teburd) [reply tests/boards/intel_adsp/hda/src/smoke.c:180]: i've added tests to cover the NUM_CPUS=1 or the platform default to deal with this, as well as wrapping all the cache flush/invalidate bits in ifdefry

File: tests/boards/intel_adsp/hda/src/smoke.c (hunk 18 above)
Comment (teburd) [reply tests/boards/intel_adsp/hda/src/smoke.c]: Fixed

File: soc/xtensa/intel_adsp/common/include/cavs_test.h (hunk 7 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/common/include/cavs_test.h]: Done

File: soc/xtensa/intel_adsp/tools/cavstool.py (hunk 8 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/tools/cavstool.py:33]: I'm happy to revisit this idea in a cleanup round, this patchset I feel like should focus on providing the basics of HDA at this point.

File: soc/xtensa/intel_adsp/tools/cavstool.py (hunk 14 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/tools/cavstool.py:568]: Todo in a cleanup round, when I move things to use HDAStream perhaps

**Comment by teburd** [comment]
Rebased without changes

**Comment by teburd** [comment]
Rebased with minor string changes (assert wording was slightly broken) and doc comments on cavs_hda_unused to address previous review comments.

File: soc/xtensa/intel_adsp/common/include/cavs_hda.h [hunk 23]
```diff
@@ -0,0 +1,298 @@
...
+#endif
+
+/* The read/write positions are masked to 24 bits */
+#define HDA_RWP_MASK 0x00FFFFFF
+
+/* Buffers must be 128 byte aligned, this mask enforces that */
+#define HDA_ALIGN_MASK 0xFFFFFF80
+
+
+#define HDA_ADDR(base, stream) (base + stream*HDA_REGBLOCK_SIZE)
```
Comment (ceolin) [inline soc/xtensa/intel_adsp/common/include/cavs_hda.h:44]: `((base) + (stream)*HDA_REGBLOCK_SIZE)` ? It helps avoid wrong behavior if someone calls it `HDA_ADDR(base, stream + offset)`

**Review by ceolin — DISMISSED** [review]
Awesome PR !

**Review by ceolin — DISMISSED** [review]

**Review by andyross — DISMISSED** [review]
Refresh +1

**Comment by mbolivar-nordic** [comment]
needs rebase?

File: dts/xtensa/intel/intel_cavs.dtsi [hunk 24]
```diff
@@ -31,5 +31,25 @@
 
 			status = "okay";
 		};
+
+		hda_host_out: dma@0x72800 {
```
Comment (abonislawski) [inline dts/xtensa/intel/intel_cavs.dtsi:35]: Probably just @72800?
Warning (simple_bus_reg): /soc/dma@0x72800: simple-bus unit address format error, expected "72800"

File: soc/xtensa/intel_adsp/common/include/cavs_hda.h [hunk 25]
```diff
@@ -0,0 +1,303 @@
...
+#define DGBRP(base, stream) ((volatile uint32_t *)(HDA_ADDR(base, stream) + 0x10))
+
+/* Gateway Buffer Write Position */
+#define DGBWP(base, stream) ((volatile uint32_t *)(HDA_ADDR(base, stream) + 0x14))
+
+/* Gateway Buffer Segment Position */
+#define DGBSP(base, stream) ((volatile uint32_t *)(HDA_ADDR(base, stream) + 0x18))
+
+/* Gateway Minimum Buffer Size */
+#define DGMBS(base, stream) ((volatile uint32_t *)(HDA_ADDR(base, stream) + 0x1c))
```
Comment (abonislawski) [inline soc/xtensa/intel_adsp/common/include/cavs_hda.h:83]: I was checking registry values in real audio playback scenario and see no use of DGMBS, this one is the only difference (SOF will put buffer size here).
Not a big deal, probably doesnt matter, just FYI

**Review by abonislawski — COMMENTED** [review]
I can confirm that it works great with SOF audio

**Review by nashif — APPROVED** [review]

**Review by lgirdwood — APPROVED** [review]

File: soc/xtensa/intel_adsp/common/include/cavs_hda.h [hunk 26]
```diff
@@ -0,0 +1,303 @@
...
+
+#define HDA_ADDR(base, stream) ((base) + (stream)*HDA_REGBLOCK_SIZE)
+
+
+/* Gateway Control and Status Register */
+#define DGCS(base, stream) ((volatile uint32_t *)HDA_ADDR(base, stream))
+#define DGCS_SCS BIT(31) /* Sample container size */
+#define DGCS_GEN BIT(26) /* Gateway Enable */
+#define DGCS_L1ETP BIT(25) /* L1 Enter Prevent */
+#define DGCS_L1EXP BIT(25) /* L1 Exit Prevent */
```
Comment (marc-hb) [inline soc/xtensa/intel_adsp/common/include/cavs_hda.h:52]: Fixed in
- #58922
